### PR TITLE
use panel for workchain and advanced setting

### PR DESCRIPTION
**Pull request by superstar54**
- use panel for the `workchain_setting` and the `advanced_setting`
- pass `parameters` to the builder, and let the plugin developer decide how to use the parameters.
- add test for configuration step. I skipped the test for set_configuration_parameters because it needs changes in code in the magnetization

### CODE REVIEW

File: src/aiidalab_qe/app/configuration/__init__.py [hunk 1]
```diff
@@ -61,8 +59,10 @@ def __init__(self, **kwargs):
         self.tab.set_title(0, "Workflow")
         self.tab.set_title(1, "Advanced settings")
 
-        # add plugin specific settings
-        self.settings = {}
+        self.settings = {
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/__init__.py:62]: Add a comment on what these settings are and why you need this variable.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 2]
```diff
@@ -336,6 +343,8 @@ def _update_widget(self, change):
     def get_magnetization(self):
         """Method to generate the dictionary with the initial magnetic moments"""
         magnetization = {}
+        if self.kinds is None:
+            return magnetization
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:347]: What is this change for, I guess it is specific to the magnetization setting widget? If not needed, please revert and open an issue and solve it separately.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 3]
```diff
@@ -344,7 +353,7 @@ def _set_magnetization_values(self, **kwargs):
         """Update used for conftest setting all magnetization to a value"""
         # self.override.value = True
         with self.hold_trait_notifications():
-            if "initial_magnetic_moments" in kwargs:
+            if "initial_magnetic_moments" in kwargs and self.kinds is not None:
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:356]: Same here.

File: src/aiidalab_qe/workflows/__init__.py [hunk 4]
```diff
@@ -73,7 +73,7 @@ def define(cls, spec):
             spec.expose_inputs(
                 plugin_workchain,
                 namespace=name,
-                exclude=entry_point["exclude"],
+                exclude=entry_point.get("exclude", []),
```
Comment (unkcpz) [inline src/aiidalab_qe/workflows/__init__.py:76]:
```suggestion
                exclude=entry_point.get("exclude"),
```
I think the default value of exclude is `None`, no? It is not good to use mutable values in a function constructor.

File: src/aiidalab_qe/workflows/__init__.py [hunk 5]
```diff
@@ -121,48 +121,48 @@ def get_builder_from_protocol(
...
         if properties is None:
             properties = []
         builder.properties = orm.List(list=properties)
         # add plugin workchain
-        codes = {"pw_code": pw_code, "dos_code": dos_code, "projwfc_code": projwfc_code}
         for name, entry_point in plugin_entries.items():
             if name in properties:
                 plugin_builder = entry_point["get_builder"](
-                    codes, structure, overrides, protocol, **kwargs
+                    codes, structure, copy.deepcopy(parameters), **kwargs
```
Comment (unkcpz) [inline src/aiidalab_qe/workflows/__init__.py:165]: Are you changing the entry_point interface? Need 4 previously but here it is 3.

**Review by unkcpz — CHANGES_REQUESTED** [review]
Some minor requests.

File: src/aiidalab_qe/app/configuration/__init__.py (hunk 1 above)
Comment (superstar54) [reply src/aiidalab_qe/app/configuration/__init__.py:65]: Done

File: src/aiidalab_qe/workflows/__init__.py (hunk 5 above)
Comment (superstar54) [reply src/aiidalab_qe/workflows/__init__.py:165]: Yes, the interface changes.

File: src/aiidalab_qe/workflows/__init__.py (hunk 4 above)
Comment (superstar54) [reply src/aiidalab_qe/workflows/__init__.py]: I checked the sources:
```
exclude: Optional[Sequence[[str]]] = None
```
The default one is `None`, but passing `[]` as a variable (note, here is not a constructor) is also safe. I used `None` as you suggested.

File: src/aiidalab_qe/app/configuration/advanced.py (hunk 2 above)
Comment (superstar54) [reply src/aiidalab_qe/app/configuration/advanced.py]: I added this when trying to add a test for the `advanced setting`. I decided not to add the test for the moment. I reverted it as you suggested.

File: src/aiidalab_qe/app/configuration/advanced.py (hunk 3 above)
Comment (superstar54) [reply src/aiidalab_qe/app/configuration/advanced.py]: I reverted it.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 6]
```diff
@@ -212,27 +215,31 @@ def get_setting_parameters(self):
...
-            self.kpoints.distance.value = parameters["kpoints_distance_override"]
-            self.kpoints.override.value = True
-        if parameters.get("degauss_override", None) is not None:
-            self.smearing.degauss.value = parameters["degauss_override"]
-            self.smearing.override.value = True
-        if parameters.get("smearing_override", None) is not None:
-            self.smearing.smearing.value = parameters["smearing_override"]
-            self.smearing.override.value = True
+    def set_panel_value(self, parameters):
+        if "pseudo_family" in parameters:
```
[the remark this answers is not in the document]
Comment (superstar54) [reply src/aiidalab_qe/app/configuration/advanced.py:231]: Yes.

File: src/aiidalab_qe/app/configuration/__init__.py [hunk 7]
```diff
@@ -108,25 +108,18 @@ def _observe_previous_step_state(self, change):
...
 
         with self.hold_trait_notifications():
-            # Work chain settings
-            workchain_settings = parameters.get("workchain_settings", {})
-            self.workchain_settings.set_setting_parameters(workchain_settings)
-            # Advanced settings
-            advanced_settings = parameters.get("advanced_settings", {})
-            self.advanced_settings.set_setting_parameters(advanced_settings)
+            for name, settings in self.settings.items():
+                if parameters.get(name, False):
```
[the remark this answers is not in the document]
Comment (superstar54) [reply src/aiidalab_qe/app/configuration/__init__.py]: I changed the `name` to `identifier`.

**Pull request merged**